modules/indexing/src/main/java/org/apache/ignite/internal/processors/query/h2/disk/ExternalResultHashIndex.java

```diff
@@ -0,0 +1,469 @@
+     * @return Byte buffer with data.
+     */
+    private Entry readEntryFromIndexFile(long slot) {
+        return readEntryFromIndexFile(slot, fileIo);
+    }
+
+    /**
+     * Reads hash map entry from the underlying file.
+     *
+     * @param slot number of the entry to read. {@code Null} means sequential reading.
+     * @param fileCh File to read from.
+     * @return Byte buffer with data.
+     */
+    private Entry readEntryFromIndexFile(Long slot, FileIO fileCh) {
+        try {
+            if (slot != null)
+                gotoSlot(slot);
+            else
+                slot = fileCh.position() / Entry.ENTRY_BYTES;
+
+            reusableBuff.clear();
+
+            while (reusableBuff.hasRemaining()) {
+                int bytesRead = fileCh.read(reusableBuff);
+
+                if (bytesRead <= 0)
+                    throw new IOException("Can not read data from file: " + idxFile.getAbsolutePath());
+            }
+
+            reusableBuff.flip();
+
+            int hashCode = reusableBuff.getInt();
+
+            // We store addr value incremented by 1 to distinguish 0 address from the empty address field.
+            // On the disk 0 means empty, 1 means zero, y means y-1.
+            long addr = reusableBuff.getLong() - 1;
+
+            return new Entry(hashCode, addr, slot);
+        }
+        catch (IOException e) {
+            U.closeQuiet(this);
+
+            throw new IgniteException("Failed to read query result the from spill idx file. [slot=" + slot + ']', e);
+        }
+    }
+
+    /**
+     * Writes hash map entry to file.
+     *
+     * @param slot Entry slot.
+     * @param hashCode Hash code.
+     * @param addr Row address.
+     */
+    private void writeEntryToIndexFile(long slot, int hashCode, long addr) {
+        try {
+            gotoSlot(slot);
+
+            reusableBuff.clear();
+            reusableBuff.putInt(hashCode);
+
+            // We store addr value incremented by 1 to distinguish 0 address from the empty address field.
+            // On the disk 0 means empty, 1 means zero, y means y-1.
+            reusableBuff.putLong(addr + 1);
+            reusableBuff.flip();
+
+            while (reusableBuff.hasRemaining()) {
+                int bytesWritten = fileIo.write(reusableBuff);
+
+                if (bytesWritten <= 0)
+                    throw new IOException("Can not write data to file: " + idxFile.getAbsolutePath());
+            }
+        }
+        catch (IOException e) {
+            U.closeQuiet(this);
+
+            throw new IgniteException("Failed to write intermediate query result to the spill file.", e);
+        }
+    }
+
+    /**
+     * Sets arbitrary file position.
+     *
+     * @param slot Position to set.
+     */
+    private void gotoSlot(long slot) {
+        try {
+            fileIo.position(slot * Entry.ENTRY_BYTES);
+        }
+        catch (Exception e) {
+            U.closeQuiet(fileIo);
```

Comment:
Do we really need to close fileIO here?

Reply:
Your idea is not clear for me. Could you please explain it in more details?